src/jit/namedintrinsiclist.h

```diff
@@ -16,10 +16,84 @@ enum NamedIntrinsic : unsigned int
     NI_System_Collections_Generic_EqualityComparer_get_Default = 4,
 #if FEATURE_HW_INTRINSICS
     NI_HW_INTRINSIC_START,
-#define HARDWARE_INTRINSIC(id, name, isa) NI_##id,
+#define HARDWARE_INTRINSIC(id, name, isa, ival, size, numarg, t1, t2, t3, t4, t5, t6, t7, t8, t9, t10, category)       \
+    NI_##id,
 #include "hwintrinsiclistxarch.h"
     NI_HW_INTRINSIC_END
 #endif
 };
 
+#if FEATURE_HW_INTRINSICS && defined(_TARGET_XARCH_)
+enum HWIntrinsicCategory : unsigned int
+{
+    // Simple SIMD intrinsics
+    // - take Vector128/256<T> parameters
+    // - return a Vector128/256<T>
+    // - generate single instruction
+    // - the codegen of overloads can be determined by intrinsicID and base type of returned vector
+    HW_Category_SimpleSIMD = 0x1,
+
+    // Scalar intrinsics
+    // - operate over general purpose registers, like crc32, lzcnt, popcnt, etc.
+    HW_Category_Scalar = 0x2,
+
+    // 64bit:
+    // - the generated instruction may require 64-bit general purpose register (r64)
+    // - must throw PlatforNotSupportedException on 32-bit platform
+    HW_Category_64bit       = 0x4,
+    HW_Category_64bitScalar = 0x6,
+
+    // IMM intrinsics
+    // - the generated instruction requires immediate value (i.e. imm8)
+    HW_Category_IMM = 0x8,
```

Comment:
This feels like it should be a flag that is used in combination with the `ival` field

Reply:
We need this flag for ordinary imm intrinsics, such checking type arguments.

Reply:
We have `HW_Category_FixedIMM`.

Reply:
Yes, but we don't need it. FixedImm can currently be detected by an ival that is not -1. We need a flag to detect the case of no imm vs op3 imm (since both are -1 for the ival in the table)

Reply:
Yes, but we don't need it. FixedImm can currently be detected by an ival that is not -1. We need a flag to detect the case of no imm vs op3 imm (since both are -1 for the ival in the table)

Reply:
> be detected by an ival that is not -1

Good point.

>  op3 imm 

There are some imm intrinsics that imm parameter is not op3.

Reply:
> There are some imm intrinsics that imm parameter is not op3

Whichever operand it is then 😄 (maybe some mechanism for identifying which operand is the immediate value would be useful, I think we can probably wait and see for that though).